x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java

```diff
@@ -313,9 +322,48 @@ private void markJobAsDeletingIfNotUsed(String jobId, ActionListener<Boolean> li
                                 + datafeedIds.iterator().next() + "] refers to it"));
                         return;
                     }
-                    jobConfigProvider.markJobAsDeleting(jobId, listener);
+                    cancelResetTaskIfExists(jobId, ActionListener.wrap(
+                        response -> jobConfigProvider.updateJobBlockReason(jobId, new Blocked(Blocked.Reason.DELETE, taskId), listener),
+                        listener::onFailure
+                    ));
                 },
                 listener::onFailure
         ));
     }
+
+    private void cancelResetTaskIfExists(String jobId, ActionListener<Boolean> listener) {
+        ActionListener<ListTasksResponse> listTasksListener = ActionListener.wrap(
+            listTasksResponse -> {
+                List<TaskInfo> tasks = listTasksResponse.getTasks();
+                if (tasks.size() == 1) {
+                    logger.info("[{}] Cancelling reset task because delete was requested", jobId);
+                    TaskInfo resetTaskInfo = tasks.get(0);
+                    CancelTasksRequest cancelTasksRequest = new CancelTasksRequest();
+                    cancelTasksRequest.setReason("deleting job");
+                    cancelTasksRequest.setActions(ResetJobAction.NAME);
+                    cancelTasksRequest.setTaskId(resetTaskInfo.getTaskId());
+                    executeAsyncWithOrigin(client, ML_ORIGIN, CancelTasksAction.INSTANCE, cancelTasksRequest, ActionListener.wrap(
+                        cancelTasksResponse -> listener.onResponse(true),
+                        e -> {
+                            if (ExceptionsHelper.unwrapCause(e) instanceof ResourceNotFoundException) {
+                                listener.onResponse(true);
+                            } else {
+                                listener.onFailure(e);
+                            }
+                        }
+                    ));
+
+                } else {
+                    listener.onResponse(false);
+                }
+            },
+            listener::onFailure
+        );
+
+        ListTasksRequest listTasksRequest = new ListTasksRequest();
+        listTasksRequest.setActions(ResetJobAction.NAME);
+        listTasksRequest.setDescriptions(MlTasks.JOB_TASK_ID_PREFIX + jobId);
+        listTasksRequest.setDetailed(true);
+        executeAsyncWithOrigin(client, ML_ORIGIN, ListTasksAction.INSTANCE, listTasksRequest, listTasksListener);
```

Comment:
Now that we have the `task_id` in the `blocked` object, is this still necessary?